pythia/models/butd.py

```diff
@@ -0,0 +1,294 @@
+        )
+        seqs = data["texts"]
+        timesteps = self.text_processor.max_length
+
+        # Lists to store completed sequences and scores
+        complete_seqs = list()
+        complete_seqs_scores = list()
+
+        # Dummy output for loss calculation to not complain
+        outputs = torch.ones(sample_list.get_batch_size(), 52, self.vocab_size).to("cuda")
+
+        for t in range(timesteps):
+            data["texts"] = self.word_embedding(data["texts"].long())
+            if "state" not in data:
+                data["state"] = None
+
+            scores, data["state"] = self.decoder(
+                data["image_features"],
+                data["texts"][:, 0, :],
+                data["state"],
+            )
+
+            # Add predicted scores to top_k_scores
+            scores = torch.nn.functional.log_softmax(scores, dim=1)
+            scores = top_k_scores.expand_as(scores) + scores
+
+            # Find next top k scores and words
+            if t == 0:
+                top_k_scores, top_k_words = scores[0].topk(k, 0, True, True)
+            else:
+                top_k_scores, top_k_words = scores.view(-1).topk(
+                    k, 0, True, True
+                )
+
+            # Convert to vocab indices
+            prev_word_inds = top_k_words / self.vocab_size
+            next_word_inds = top_k_words % self.vocab_size
+
+            # Add new words to sequences
+            seqs = torch.cat(
+                [seqs[prev_word_inds], next_word_inds.unsqueeze(1)], dim=1
+            )
+
+            # Find completed sequences
+            incomplete_inds = [
+                ind
+                for ind, next_word in enumerate(next_word_inds)
+                if next_word != self.vocab.EOS_INDEX
+            ]
+            complete_inds = list(set(range(len(next_word_inds))) - set(incomplete_inds))
+
+            # Add to completed sequences
+            if len(complete_inds) > 0:
+                complete_seqs.extend(seqs[complete_inds].tolist())
+                complete_seqs_scores.extend(top_k_scores[complete_inds])
+
+            # Reduce beam length
+            k -= len(complete_inds)
+
+            # Proceed with incomplete sequences
+            if k == 0:
+                break
+            seqs = seqs[incomplete_inds]
+
+            data = self.get_data_beam_search_t(data, prev_word_inds[incomplete_inds])
+            top_k_scores = top_k_scores[incomplete_inds].unsqueeze(1)
+            data["texts"] = next_word_inds[incomplete_inds].unsqueeze(1)
+
+
+        if len(complete_seqs_scores) == 0:
+            captions = torch.FloatTensor([0] * timesteps).unsqueeze(0)
+        else:
+            i = complete_seqs_scores.index(max(complete_seqs_scores))
+            captions = torch.FloatTensor(complete_seqs[i]).unsqueeze(0)
+
+        targets = sample_list.answers[:, 0]
+        sample_list.add_field("targets", targets)
+        sample_list.add_field("captions", captions)
```

Comment:
`sample_list` requires a `targets` field otherwise it doesn't calculate loss. The `targets` is generated after we sort the captions in this batch in decreasing order. That is why I added it here. I can move the `captions` to `model_output`.